src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,122 @@
+     * @throws IncompleteInputException If the command given is incomplete.
+     * @throws WrongMarkException       If the command given is wrong.
+     * @throws InvalidInputException    If the command given is invalid.
+     */
+    public static boolean checkCommandType(String commandGiven,
+                                           TaskList tasks,
+                                           Ui ui,
+                                           Storage storage)
+            throws IncompleteInputException,
+            WrongMarkException,
+            InvalidInputException {
+        commandGiven = commandGiven.trim();
+        String[] splittedCommands = commandGiven.split(" ");
+        String commandType = splittedCommands[0];
+        switch (commandType) {
+        case "list":
+            ui.separatorLines();
+            tasks.printTasks();
+            break;
+        case "mark":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                if (Objects.equals(task.getStatusIcon(), "X")) {
+                    ui.showAlreadyDone();
+                } else {
+                    ui.separatorLines();
+                    task.setAsDone(task);
+                }
+                storage.save(tasks);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                ui.showInvalidIndex();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            break;
+        case "unmark":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                if (Objects.equals(task.getStatusIcon(), " ")) {
+                    ui.showAlreadyUndone();
+                } else {
+                    ui.separatorLines();
+                    task.setAsUndone(task);
+                }
+                storage.save(tasks);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                ui.showInvalidIndex();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            break;
+        case "delete":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                tasks.deleteTask(Integer.parseInt(splittedCommands[1]));
+                ui.deleteTaskMessage(task);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                ui.showInvalidIndex();
+            }
+            break;
+        case "find":
+            ui.separatorLines();
+            try {
+                String keyword = splittedCommands[1];
+                tasks.findTasks(keyword);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                throw new IncompleteInputException("find what eh! ");
+            }
+            break;
+        case "bye":
+            try {
+                storage.save(tasks);
```

Comment:
Redundant line. It's not necessary to save. Tasks should be saved after every CRUD command is run.

Reply:
I thought to be on safer side by adding save option here as well. In case of any file handling error.